src/integration-test/java/com/commercetools/sync/integration/commons/utils/ProductITUtils.java

```diff
@@ -48,19 +48,68 @@ public static void deleteProductSyncTestData(@Nonnull final SphereClient ctpClie
     }
 
     /**
-     * Deletes all products from the CTP project defined by the {@code ctpClient}.
+     * Unpublishes all published products, then deletes all products from the CTP project defined by the
+     * {@code ctpClient}.
      *
-     * @param ctpClient defines the CTP project to delete the categories from.
+     * @param ctpClient defines the CTP project to delete the products from.
      */
     public static void deleteAllProducts(@Nonnull final SphereClient ctpClient) {
-        final Function<Product, SphereRequest<Product>> unpublishAndDelete = product -> {
-            if (product.getMasterData().isPublished()) {
-                product = ctpClient.execute(ProductUpdateCommand.of(product, Unpublish.of()))
-                                   .toCompletableFuture().join();
-            }
-            return ProductDeleteCommand.of(product);
-        };
-        queryAndApply(ctpClient, ProductQuery::of, unpublishAndDelete);
+        queryAndCompose(ctpClient, ProductQuery::of, product -> deleteProduct(ctpClient, product));
+    }
+
+    /**
+     * If the {@code product} is published, issues an unpublish request followed by a delete. Otherwise,
+     * issues a delete request right away.
+     *
+     * @param ctpClient defines the CTP project to delete the product from.
+     * @param product the product to be deleted.
+     * @return a {@link CompletionStage} containing the deleted product.
+     */
+    @Nonnull
+    private static CompletionStage<Product> deleteProduct(@Nonnull final SphereClient ctpClient,
```

Comment:
`safeDeleteProduct` then below function can be called `unpublishAndDeleteProduct` and `deleteProduct`

Reply:
34e4470